x/gamm/keeper/swap.go

```diff
@@ -31,6 +31,11 @@ func (k Keeper) SwapExactAmountIn(
 	}
 	tokensIn := sdk.Coins{tokenIn}
 
+	pool, ok := poolI.(types.TraditionalAmmInterface)
+	if !ok {
+		return sdk.Int{}, errors.New("type cast failed")
```

Comment:
consider formatting actual and expected types and also covering this with a test

Reply:
Im not quite sure how to go about formatting the error to show an "expected" vs "actual" type without just hard coding it.

Reply:
I'd do something like

```
 fmt.Errorf("failed cast to TraditionalAmmInterface, actual type: %T, poolI)
```

Reply:
but not a blocker, feel free to skip if you want

Reply:
ah, duh, thanks 😅